kani-driver/src/concrete_playback/test_generator.rs

```diff
@@ -84,71 +94,83 @@ impl KaniSession {
+
+        if is_new_injection {
+            let src_path = Path::new(src_path);
+            let (path, file_name) = extract_parent_dir_and_src_file(src_path)?;
+            let file_line_ranges = vec![FileLineRange {
+                file: file_name,
+                line_range: Some((unit_test_start_line, unit_test_end_line)),
+            }];
+            self.run_rustfmt(&file_line_ranges, Some(&path))?;
         }
 
-        // Run rustfmt on just the inserted lines.
-        let concrete_playback_num_lines = unit_test.code.len();
-        let unit_test_start_line = proof_harness_end_line + 1;
-        let unit_test_end_line = unit_test_start_line + concrete_playback_num_lines - 1;
-        let src_path = Path::new(src_path);
-        let (path, file_name) = extract_parent_dir_and_src_file(src_path)?;
-        let file_line_ranges = vec![FileLineRange {
-            file: file_name,
-            line_range: Some((unit_test_start_line, unit_test_end_line)),
-        }];
-        self.run_rustfmt(&file_line_ranges, Some(&path))?;
         Ok(())
     }
 
     /// Writes the new source code to a user's source file using a tempfile as the means.
-    /// Returns whether the unit test was already in the old source code.
+    /// Returns whether new unit test was injected.
     fn add_test_inplace(
         &self,
         source_path: &str,
         proof_harness_end_line: usize,
-        unit_test: &UnitTest,
+        mut unit_tests: Vec<UnitTest>,
     ) -> Result<bool> {
         // Read from source
         let source_file = File::open(source_path).unwrap();
         let source_reader = BufReader::new(source_file);
+        let source_string = read_to_string(source_path).unwrap();
 
-        // Create temp file
-        let mut temp_file = TempFile::try_new("concrete_playback.tmp")?;
-        let mut curr_line_num = 0;
-
-        // Use a buffered reader/writer to generate the unit test line by line
-        for line in source_reader.lines().flatten() {
-            if line.contains(&unit_test.name) {
+        // filter out existing harnesses.
+        unit_tests.retain(|unit_test| {
+            if source_string.contains(&unit_test.name) {
                 if !self.args.common_args.quiet {
                     println!(
                         "Concrete playback unit test `{}/{}` already found in source code, so skipping modification.",
                         source_path, unit_test.name,
                     );
                 }
-                // the drop impl will take care of flushing and resetting
-                return Ok(true);
+
+                false
+            } else {
+                true
             }
+        });
+
+        // Create temp file
+        let mut temp_file = TempFile::try_new("concrete_playback.tmp")?;
+        let mut curr_line_num = 0;
+
+        // Use a buffered reader/writer to generate the unit test line by line
+        let mut is_new_injection = false;
```

Comment:
Is this needed? I'm assuming this will be true if `!unit_tests.is_empty()`.

Reply:
930efaf7f4edfaa107774da0e770ce8fe252b931